nighthawk/test/statistic_test.cc

```diff
@@ -0,0 +1,199 @@
+  // + 13, 109 + 16). Again the estimated population variance of 30 is computed correctly by the
+  // two-pass algorithm, but the naïve algorithm now computes it as −170.66666666666666. This is a
+  // serious problem with naïve algorithm and is due to catastrophic cancellation in the subtraction
+  // of two similar numbers at the final stage of the algorithm.
+  std::vector<int> values{4, 7, 13, 16};
+  int exponential = 0;
+  for (exponential = 3; exponential < 16; exponential++) {
+    TypeParam a;
+    double offset = std::pow(10, exponential);
+    for (int value : values) {
+      a.addValue(offset + value);
+    }
+    // If an implementation makes this claim, we put it to the test. SimpleStatistic is simple and
+    // fast, but starts failing this test when exponential equals 8. HdrStatistic breaks at 5.
+    // TODO(oschaaf): evaluate ^^
+    if (a.resistsCatastrophicCancellation()) {
+      Helper::expectNear(22.5, a.pvariance(), a.significantDigits());
+      Helper::expectNear(4.7434164902525691, a.pstdev(), a.significantDigits());
+    }
+  }
+}
+
+TYPED_TEST(TypedStatisticTest, OneMillionRandomSamples) {
+  std::random_device rd;
+  std::mt19937 mt(rd());
+  // TODO(oschaaf): Actually the range we want to test is a factor 1000 higher, but
+  // then catastrophical cancellation make SimpleStatistic fail expectations.
+  // For now, we use values that shouldn't trigger the phenomena. Revisit this later.
+  std::uniform_real_distribution<double> dist(1ULL, 1000ULL * 1000 * 60);
+
+  StreamingStatistic referenceStatistic;
+  TypeParam testStatistic;
+  for (int i = 0; i < 999999; ++i) {
+    auto value = dist(mt);
+    referenceStatistic.addValue(value);
+    testStatistic.addValue(value);
+  }
+  Helper::expectNear(referenceStatistic.mean(), testStatistic.mean(),
+                     testStatistic.significantDigits());
+  Helper::expectNear(referenceStatistic.pvariance(), testStatistic.pvariance(),
+                     testStatistic.significantDigits());
+  Helper::expectNear(referenceStatistic.pstdev(), testStatistic.pstdev(),
+                     testStatistic.significantDigits());
+}
+
+TYPED_TEST(TypedStatisticTest, ProtoOutput) {
+  TypeParam a;
+
+  a.addValue(6543456);
+  a.addValue(342335);
+
+  const nighthawk::client::Statistic proto = a.toProto();
+
+  EXPECT_EQ(proto.count(), 2);
+  EXPECT_EQ(proto.mean().nanos(), std::round(a.mean()));
+  EXPECT_EQ(proto.pstdev().nanos(), std::round(a.pstdev()));
+}
+
+class StatisticTest : public testing::Test {};
+
+TEST(StatisticTest, HdrStatisticPercentilesProto) {
+  HdrStatistic statistic;
+  int max = 10;
+
+  for (int i = 1; i <= max; i++) {
+    statistic.addValue(i);
+  }
+
+  std::string str;
+  google::protobuf::util::JsonPrintOptions options;
+  google::protobuf::util::MessageToJsonString(statistic.toProto(), &str, options);
+  std::ifstream myfile;
+  myfile.open("nighthawk/test/hdr_proto_json.gold");
```

Comment:
Take a look at `TestEnvironment::runfilesPath()` in the main Envoy repo; can we use this here? Or some similar abstraction? The main concern I have is that when we import into Google and make use of this internally, this path calculation changes.